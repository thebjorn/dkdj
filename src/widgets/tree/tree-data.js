

import dk from "../../dk-obj";
import Class from "../../lifecycle/coldboot/dk-class";
import counter from "../../core/counter";
import {json} from "../../browser/dk-client";

export class Leaf extends Class {
    constructor(...args) {
        super({
            id: null,
            kind: '',
            name: '',
            parent: null,
            path: [],
            root: undefined,
            children: [],
        }, ...args);
    }

    visit(location, ctx) {
        return ctx.leaf(location, this);
    }
}


export class Tree extends Leaf {
    visit(location, ctx) {
        return ctx.tree(location, this);
    }
}


export class Data extends Class {
    constructor(...args) {
        super({
            depth: 0,           // int
            height: 0,          // int
            roots: [],          // [id1,... idk]
            cache: {},          // id -> tree|leaf
        }, ...args);
        if (this.roots) this.map_objects(this);
    }
    
    map_objects({depth, height, roots, cache}) {
        if (Object.values(cache).every(v => v instanceof Tree || v instanceof Leaf)) return;
        
        this.depth = depth;
        this.height = height;
        this.roots = roots;
        this.cache = cache;
        
        const is_leaf = node => node.children.length === 0 && this.roots.indexOf(node.id) === -1;

        // convert all items in cache to Tree|Leaf objects
        Object.values(this.cache).forEach(node => {
            this.cache[node.id] = new (is_leaf(node) ? Leaf : Tree)(node);
        });

        // convert all child-ids to child objects.
        Object.values(this.cache).forEach(node => {
            node.children = node.children.map(id => this.cache[id]);
        });

        // convert all root-ids to objects.
        this.roots = this.roots.map(rootid => this.cache[rootid]);
    }
}


/*
 *  A Forrest is a tree structure with 1-n root nodes.
 *
 *  Each node _must_ have the following properties:
 *
 *    - name (the text to display in the tree)
 *    - children (a list of nodes that are this node's direct descendants).
 *
 *  a node _may_ in addition have
 *
 *    - kind (node type, a string, default will be set to 'tree' or 'leaf')
 *    - id (this should be a string to avoid surprises due to ordering issues
 *          in javascript, by default this will be set to an internally
 *          semi-random number).
 *
 */
export class Forrest {
    constructor(rootnodes) {
        this.depth = undefined;
        this.height = undefined;
        this.roots = null;
        this.cache = undefined;

        const self = this;
        this.cache = {};

        const iter = function iter(nodes, fn, parent) {
            if (!nodes) return;
            nodes.forEach(function (node) {
                fn(node, parent);
                iter(node._children, fn, node);
            });
        };

        // identify (id), categorize (kind), and linearize (cache)
        iter(rootnodes, function (node) {
            if (!node.id) node.id = counter('id');
            if (!node.kind) node.kind = node._children ? 'tree' : 'leaf';
            self.cache[node.id] = node;
        });

        // parent and path
        iter(rootnodes, function (node, parent) {
            node.parent = parent || null;
            if (node.parent) {
                node.path = node.parent.path.concat([node.parent.id]);
            } else {
                node.path = [];
            }
        });

        // childids
        iter(rootnodes, function (node) {
            node.children = (node._children || []).map(function (n) { return n.id; });
        });

        // root node
        rootnodes.forEach(function (r) {
            r.root = r.id;
            iter(r._children, function (node) {
                node.root = r.id;
            });
        });

        this.depth = Math.max.apply(null, this.cachevals().map(function (v) {
            return 1 + v.path.length;
        }));

        this.height = Math.max.apply(null, this.cachevals().map(function (v) {
            return v.children.length;
        }));
        this.height = Math.max(rootnodes.length, this.height);

        // this.outerHeight = this.cachevals().map(function (v) {
        //     return v.children.length + 1;
        // }).reduce(function (a, b) { return a+b;}, 0);

        this.roots = rootnodes;
    }

    cachekeys() {
        return Object.keys(this.cache);
    }

    cachevals() {
        const self = this;
        return this.cachekeys().map(function (key) {
            return self.cache[key];
        });
    }
}


export class DataSource extends Data {
    fetch() {
        dk.trigger(this, 'fetch-data-start', this);
        dk.trigger(this, 'fetch-data', this);
    }
}


export class JSonDataSource extends DataSource {
    constructor(...args) {
        super({
            cache: {},
            roots: [],
            depth: 0,
            height: 0,
        }, ...args);
        this.__fetched = false;
        this.fetch();
    }
    
    fetch() {
        if (this.__fetched) return;
        dk.trigger(this, 'fetch-data-start', this);
        this.map_objects(this);
        this.__fetched = true;
        dk.trigger(this, 'fetch-data', this);
        return this;
    }
}


export class AjaxDataSource extends JSonDataSource {
    constructor(...args) {
        super({
            url: '',
        }, ...args);
        this.__in_json = false;
    }
    
    /*
     *  Fetch data from url.
     */
    fetch() {
        if (this.__in_json) return;
        this.__in_json = true;

        const self = this;
        dk.trigger(this, 'fetch-data-start', this);
        return json({
            url: this.url,
            cache: true,
            success: function (data) {
                self.map_objects(data);
                self.__in_json = false;
                dk.trigger(self, 'fetch-data', self);
            }
        });
    }
}


const tree_data = {
    Leaf,
    Tree,
    Data,
    Forrest,
    DataSource,
    JSonDataSource,
    AjaxDataSource
};

export default tree_data;
