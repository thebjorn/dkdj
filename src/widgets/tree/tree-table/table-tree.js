
import {UIWidget} from "../../../widgetcore/ui-widget";
import {where} from "../../../collections";
import {Generation} from "./generation";

/*
 *  This is the top-most widget.
 *
 *  The get_selected() method returns the list of objects that are currently
 *  selected.  One or more objects can be selected depending on the
 *  multiselect attribute (default true).
 *
 *  The direct children of this widget are `dk.tree.Generation` objects
 *  that represent the panels of this widget. The children are created
 *  in the `draw()` method.
 *
 */

export class TableTree extends UIWidget {
    constructor(...args) {
        super({
            type: 'table-tree',
            title: 'MISSING-TITLE',
            item_height: 25,
            select: 'leaves',
            //cursor_children: "url(http://static.datakortet.no/cur/29.cur), pointer",
            //cursor_select: 'url(http://static.datakortet.no/cur/71.cur), pointer',
            selected: null,
            target: null,
            selectable: [],
            multiselect: true,
            values: null,
            //_tree_cache: {},
            _generation: null,
            current_node: null,
            _selected_nodes: null,
            size: 5,

            structure: {
                classes: ['dk-tree-select'],
                css: {
                    display: 'flex',
                }
            },
        }, ...args);
        if (typeof this.selectable === 'string') {
            this.selectable = this.selectable === ''? null: this.selectable.split(',');
        }
        if (typeof this.multiselect === 'string') {
            this.multiselect = (this.multiselect === '1');
        }
        this.values = {};
        this._generation = [];
        this._selected_nodes = [];
    }
    
    construct() {
        this.widget().height(this.size * this.item_height);
    }

    get_selected() {
        if (! (this.data || this.data.cache)) return {};
        return where(this.data.cache, {selected: true});
    }

    selection_state(node, boolval) {
        if (boolval && !this.multiselect) {
            // if we're selecting a new node in single-select mode,
            // remove all other selections.
            this._selected_nodes.forEach(function (snode) {
                snode.selection_state(false);
            });
            this._selected_nodes.length = 0;  // clear array
        }
        if (boolval && !this._selected_nodes.includes(node)) {
            this._selected_nodes.push(node);
        }
        if (!boolval) {
            const _pos = this._selected_nodes.indexOf(node);
            if (_pos !== -1) this._selected_nodes.splice(_pos, 1);
        }
        node.selection_state(boolval);
    }

    navigate_to(node) {
        if (this.current_node !== node) {
            // deselect current node.
            if (this.current_node !== null) {
                this.current_node.navigation_state(false);
            }
            this.current_node = node;
            node.navigation_state(true);
            node.widget().focus();
        }
    }

    draw(data) {
        const self = this;
        data = data || this.data;
        for (let i=0; i<data.depth; i++) {
            const generation = Generation.append_to(this.widget(), {
                depth: i,
                title: this.title,
                selected: null,
                width: Math.floor((self.widget().innerWidth() - 0.5) / data.depth),
                tree: self
            });
            this._generation.push(generation);
            if (i > 0) this._generation[i - 1].next = generation;
        }
        console.info("ROOTS:", this.data.roots);
        if (true || this.data.depth > 0) {
            this._generation[0].parent = this;
            this._generation[0].draw(this.data.roots);
        }
    }
}
