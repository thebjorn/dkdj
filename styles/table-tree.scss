@import "default-theme";

.TableTree {
    overflow-y: hidden;
    border: 1px solid #e0e0e0;
    box-sizing: border-box;
    
    .dk-tree-generation {
        &::-webkit-scrollbar { width: 6px; }
        &::-webkit-scrollbar-track {
            -webkit-box-shadow: inset 0 0 6px rgba(0,0,0,0.3);
        }
        &::-webkit-scrollbar-thumb {
            background-color: darkgrey;
            outline: 1px solid slategrey;
        }

        //background-color: red;
        
        direction: rtl;
        
        .treeselect-nodelist {
            direction: ltr;
            //td { border: 1px solid red; }
            tr.tree-node {
                td.treeselect-input {
                    width: 1em;
                    text-align: center;
                }
                td.treeselect-node {}
                td.treeselect-icon {
                    width: 1em;
                    text-align: right;
                }
            }
            tr.nav-selected {
                background-color: $selected-color;
            }
        }
    }
}
