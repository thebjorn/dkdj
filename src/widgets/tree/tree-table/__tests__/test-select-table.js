
import $ from 'jquery';
import {TableTree} from "../table-tree";
import page from "../../../../widgetcore/dk-page";
import utidy from "../../../../browser/dk-html";
import {Forrest, JSonDataSource} from "../../tree-data";



test("test-select-table-render", () => {
    document.body.innerHTML = `
        <div id="work">
        </div>
    `;
    const work = $('#work');
    page.initialize(document);

    const tree = TableTree.create_inside(work, {
        selectable: ['department', 'company'],
        multiselect: false,
        tree_data: JSonDataSource.create({
            "depth": 2,
            "height": 2,
            "cache": {
                "d329": {
                    "kind": "department",
                    "name": "Bar3",
                    "parent": "c57",
                    "id": "d329",
                    "path": [
                        "c57", "d329"
                    ],
                    "root": "c57",
                    "children": []
                },
                "c57": {
                    "kind": "company",
                    "name": "Bar1",
                    "parent": null,
                    "id": "c57",
                    "path": [
                        "c57"
                    ],
                    "root": "c57",
                    "children": ['d329', 'd328']
                },
                "d328": {
                    "kind": "department",
                    "name": "Bar2",
                    "parent": "c57",
                    "id": "d328",
                    "path": [
                        "c57", "d328"
                    ],
                    "root": "c57",
                    "children": []
                }
            },
            "roots": ['c57']
        })
    });
    // console.log("WORK:", work.html());
    $('[name=Bar1] .treeselect-node').click();
    const selected = tree.get_selected();
    expect(selected).toHaveLength(1);
    expect(selected[0].id).toBe("c57");
    expect(work.html()).toMatchSnapshot();
});
