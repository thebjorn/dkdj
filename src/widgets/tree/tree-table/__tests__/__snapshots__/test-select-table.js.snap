// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test-select-table-render 1`] = `
"
        <div id=\\"table-tree-1\\" class=\\"table-tree dk-tree-select\\" style=\\"display: flex; height: 125px;\\"><div id=\\"generation-1\\" class=\\"dk-tree-generation\\" style=\\"overflow-y: scroll; overflow-x: none; direction: rtl; width: 0px;\\" generation=\\"0\\"><table id=\\"node-list-1\\" class=\\"treeselect-nodelist dk-tree-nodelist table table-condensed dk-bx\\" style=\\"width: 100%;\\"><thead id=\\"dk-bx-1\\">
<tr id=\\"dk-table-1\\">
<th colspan=\\"3\\" style=\\"height: 20px;\\" id=\\"dk-table-2\\">MISSING-TITLE</th>
</tr>
</thead><tbody id=\\"dk-bx-2\\">
<tr id=\\"dk-table-3\\" class=\\"tree-node treeselect-item company treeselect-tree\\" name=\\"Bar1\\" style=\\"cursor: pointer;\\">
<td id=\\"dk-bx-3\\" class=\\"treeselect-input\\" style=\\"text-align: center; vertical-align: middle;\\"></td>

<td id=\\"dk-bx-4\\" class=\\"treeselect-node\\">Bar1</td>

<td valign=\\"middle\\" id=\\"dk-bx-5\\" class=\\"treeselect-icon\\"></td>
</tr>
</tbody></table></div><div id=\\"generation-2\\" class=\\"dk-tree-generation\\" style=\\"overflow-y: scroll; overflow-x: none; direction: rtl; width: 0px;\\" generation=\\"1\\"></div></div>"
`;
